### PR TITLE
Add Persistence

### LINKED ISSUE #6: Can I ask if the plugin data can't be persisted?
## 问题复现

保存书签后退出yazi，再次进入鸭子，丢失全部书签

## 环境

- Arch Linux
- yazi 0.2.4 (Arch Linux 2024-03-10)

### PR DESCRIPTION
**Pull request by dedukun**
closes #6 and #9

### CODE REVIEW

**Comment by dedukun** [comment]
Hey @sxyazi, can you take a look just to make sure that I'm using the `pub_static` correctly?

There's seems to be an issue, or maybe this is just how DDS is suppose to work, but this only saves to persistent memory if there's a different instance of Yazi also `sub_remote` the same time. If only the instance that is doing the `pub_remote` is subbed it's not saving the data.

**Comment by sxyazi** [comment]
Sure, it sounds like a bug, I'll investigate it

**Comment by sxyazi** [comment]
Hi, I pushed a new commit, please give it a try

**Comment by dedukun** [comment]
Hey,

It still seems to need to have two a different instance of Yazi subbed


https://github.com/dedukun/bookmarks.yazi/assets/25795432/7099cbe0-4ff2-4324-8ed2-d9372bcf1c22

**Comment by dedukun** [comment]
Oh I had a hidden terminal with Yazi running, that's why it was working... :smile:

**Comment by sxyazi** [comment]
I pushed a new commit just now, please try it again

**Comment by dedukun** [comment]
Thanks @sxyazi , I think that it is working correctly now.

File: init.lua [hunk 1]
```diff
@@ -202,8 +202,8 @@ return {
 
 		if action == "jump" then
 			ya.manager_emit("cd", { bookmarks[selected].desc })
-			ya.manager_emit("arrow", { -99999999 })
-			ya.manager_emit("arrow", { bookmarks[selected].cursor })
+			ya.manager_emit("arrow", { tostring(-99999999) })
+			ya.manager_emit("arrow", { tostring(bookmarks[selected].cursor) })
```
Comment (sxyazi) [inline init.lua:206]:
```suggestion
			ya.manager_emit("arrow", { -99999999 })
			ya.manager_emit("arrow", { bookmarks[selected].cursor })
```

The [latest commit](https://github.com/sxyazi/yazi/commit/ff14b9a265e8ef09e0313277a1b306b0fe804fde) should include fixes for it, so dont need to add `tostring`

**Comment by dedukun** [comment]
I'm not sure if this is even an issue, I can always listen to the DDS event in `cd` and only do the `arrow` then.

**Comment by sxyazi** [comment]
> The manager_emit doesn't run in sync right? Maybe it does the arrow before finishing the cd?

Yeah `manager_emit` is async, but it does ensure the order of commands.

> I can always listen to the cd DDS event and only do the arrow then.

I'm not very sure I understand what you mean. If you `cd` into a new directory (that hasn't been loaded before), triggering `arrow` at this point won't do anything because until the directory is fully loaded (or partially loaded), there are no files in that directory. `arrow` will be ignored when executed on an empty directory.

If that's the issue, currently the only solution is to use loop + `ya.sleep()`. I'll add a utility function in the next target to better address it.

**Comment by DreamMaoMao** [comment]
> ya.manager_emit("cd", { bookmarks[selected].desc })
> ya.manager_emit("arrow", { -99999999 })
> ya.manager_emit("arrow", { bookmarks[selected].cursor })

1. I think this is an incorrect logic on this plugin, the three actions are supposed to be asynchronous, but you think of them as sequential synchronous actions.

2. Second, there is no need to record the cursor, you can directly record the `file url`, and then restore by direct `cd` `file url` , do not need to `cd`to  the `cwd` and  `cursor`

**Comment by DreamMaoMao** [comment]
> Thanks, I updated to the last commit and it is working has you said.
> 
> Just a side note, I noticed that when running the jump, i.e.
> 
> ```
> ya.manager_emit("cd", { bookmarks[selected].desc })
> ya.manager_emit("arrow", { -99999999 })
> ya.manager_emit("arrow", { bookmarks[selected].cursor })
> ```
> 
> that sometimes the arrows don't do anything, I must run it twice to jump to the cursor. The `manager_emit` doesn't run in sync right? Maybe it does the `arrow` before finishing the `cd`?
> 
> This is not an issue in the last commit, I think that it was happening before.

You should not record `cwd` and `cursor`,

you  just record the `url` of the file,

Then you can `cd`  the `url` to jump to where you have record.

https://github.com/DreamMaoMao/bookmarks-persistence.yazi/blob/f974ea19282fcad57e21a0ec12bffcd4228da3ff/init.lua#L103


https://github.com/DreamMaoMao/bookmarks-persistence.yazi/blob/f974ea19282fcad57e21a0ec12bffcd4228da3ff/init.lua#L264

**Comment by sxyazi** [comment]
Ah yeah, that reminds me, you can just use the [`reveal` command](https://yazi-rs.github.io/docs/configuration/keymap/#manager.reveal), it handles all that automatically, even if there aren't any files in the directory yet. It'll hover over the correct file once the directory files are loaded.

**Comment by dedukun** [comment]
> I'm not very sure I understand what you mean. If you `cd` into a new directory (that hasn't been loaded before), triggering `arrow` at this point won't do anything because until the directory is fully loaded (or partially loaded), there are no files in that directory. `arrow` will be ignored when executed on an empty directory.
> 
> If that's the issue, currently the only solution is to use loop + `ya.sleep()`. I'll add a utility function in the next target to better address it.

Yes I think that's what was happening, I'll have a look at `reveal` has you said, I didn't know about that command.

**Comment by dedukun** [comment]
> > Thanks, I updated to the last commit and it is working has you said.
> > Just a side note, I noticed that when running the jump, i.e.
> > ```
> > ya.manager_emit("cd", { bookmarks[selected].desc })
> > ya.manager_emit("arrow", { -99999999 })
> > ya.manager_emit("arrow", { bookmarks[selected].cursor })
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > that sometimes the arrows don't do anything, I must run it twice to jump to the cursor. The `manager_emit` doesn't run in sync right? Maybe it does the `arrow` before finishing the `cd`?
> > This is not an issue in the last commit, I think that it was happening before.
> 
> You should not record `cwd` and `cursor`,
> 
> you just record the `url` of the file,
> 
> Then you can `cd` the `url` to jump to where you have record.
> 
> https://github.com/DreamMaoMao/bookmarks-persistence.yazi/blob/f974ea19282fcad57e21a0ec12bffcd4228da3ff/init.lua#L103
> 
> https://github.com/DreamMaoMao/bookmarks-persistence.yazi/blob/f974ea19282fcad57e21a0ec12bffcd4228da3ff/init.lua#L264

Oh, I also didn't know that we could use `cd` like that. Thanks I'll also have a look at it.

**Pull request merged**